server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/fastfail/SpamAssassinHandler.java

```diff
@@ -119,42 +119,41 @@ public void setSpamdRejectionHits(double spamdRejectionHits) {
-                        LOGGER.info(buffer);
-
-                        // Message reject .. abort it!
-                        return HookResult.builder()
-                            .hookReturnCode(HookReturnCode.deny())
-                            .smtpDescription(DSNStatus.getStatus(DSNStatus.PERMANENT, DSNStatus.SECURITY_OTHER)
-                                + " This message reach the spam hits treshold. Please contact the Postmaster if the email is not SPAM. Message rejected")
-                            .build();
-                    }
-                } catch (NumberFormatException e) {
-                    // hits unknown
-                }
-            }
-        } catch (MessagingException e) {
-            LOGGER.error(e.getMessage());
-        }
-        return HookResult.DECLINED;
+    public Mono<HookResult> onMessage(SMTPSession session, Mail mail) {
+        SpamAssassinInvoker sa = new SpamAssassinInvoker(metricFactory, spamdHost, spamdPort);
+
+           return  Mono.fromCallable(mail::getMessage)
+                 .flatMap(sa::scanMail)
+                 .map(result -> {
+                     // Add the headers
+                     result.getHeadersAsAttributes().forEach(mail::setAttribute);
+
+                     // Check if rejectionHits was configured
+                     if (spamdRejectionHits > 0) {
+                             double hits = Double.parseDouble(result.getHits());
+
+                             // if the hits are bigger the rejectionHits reject the
+                             // message
+                             if (spamdRejectionHits <= hits) {
+                                 String buffer = "Rejected message from " + session.getAttachment(SMTPSession.SENDER, State.Transaction).toString() + " from host " + session.getRemoteAddress().getHostName() + " (" + session.getRemoteAddress().getAddress().getHostAddress() + ") This message reach the spam hits treshold. Required rejection hits: " + spamdRejectionHits + " hits: " + hits;
+                                 LOGGER.info(buffer);
+
+                                 // Message reject .. abort it!
+                                 return HookResult.builder()
+                                     .hookReturnCode(HookReturnCode.deny())
+                                     .smtpDescription(DSNStatus.getStatus(DSNStatus.PERMANENT, DSNStatus.SECURITY_OTHER)
+                                         + " This message reach the spam hits treshold. Please contact the Postmaster if the email is not SPAM. Message rejected")
+                                     .build();
+                             }
+                     }
+                     return HookResult.DECLINED;
+                 }).onErrorResume(MessagingException.class,  e -> {
+                 LOGGER.error(e.getMessage());
```

Comment:
could we have a proper log (without suppressing the stack)?

Reply:
The indent looks weird